ggml-cuda.cu

```diff
@@ -8244,9 +8249,25 @@ static void ggml_cuda_op_mul_mat(
                         // If dst is a vector with ne0 == 1 then you don't have to do this but it still produces correct results.
                         float * dhf_dst_i = (float *) ((char *) dst_off_device + i02*nb2 + i03*nb3);
                         GGML_ASSERT(dst->nb[1] == ne0*sizeof(float));
-                        dhf_dst_i += src1_col_0*ne0 + row_low[id];
-                        CUDA_CHECK(cudaMemcpy2DAsync(dhf_dst_i, ne0*sizeof(float), dst_dd_i, row_diff*sizeof(float),
-                                                    row_diff*sizeof(float), src1_ncols, kind, stream));
+                        dhf_dst_i += src1_col_0*ne0 + dev[id].row_low;
+#if !defined(GGML_USE_HIPBLAS)
+                        if (kind == cudaMemcpyDeviceToDevice && id != g_main_device) {
+                            // cudaMemcpy2DAsync may fail with copies between vmm pools of different devices
+                            cudaMemcpy3DPeerParms p = {};
+                            p.dstDevice = g_main_device;
+                            p.dstPtr = make_cudaPitchedPtr(dhf_dst_i, ne0*sizeof(float), row_diff, src1_ncols);
```

Comment:
I think this should be:
```suggestion
                            p.dstPtr = make_cudaPitchedPtr(dhf_dst_i, ne0*sizeof(float), ne0, src1_ncols);
```
But at the same time I'm not sure that data is even used.

Reply:
It isn't used, I think it only checks that it is bigger than the extent. But I think that `row_diff` here is slightly more correct, because `dhf_dst_i` is offset by `row_low`. I guess it could be increased up to `ne0 - row_high`.

Reply:
The third argument is supposed to be "Logical width of allocation in elements". The pointer points to the buffer on the main  device that has enough space for the entire `dst` tensor of size `dst0*dst1`. Even if offset to the correct position for a partial result my interpretation of the documentation is that the value should be the pitch divided by element size. Also consider that when adding a third device the value for `row_diff` changes even though the buffer on the main device stays the same.

Reply:
I don't think that that the pitch needs to be divisible by the number of elements, the pitch is used, in some cases at least, to allow rows to have a larger alignment than each individual element. Setting it to `ne0` would mean that the last row is out of bounds of the allocated memory.

Reply:
Okay, but fundamentally we are talking about the property `dstPtr` which presumably holds information about the destination memory buffer. The row difference on the non-main device is not a property of that destination buffer. So I cannot imagine that that is the intended use.

Reply:
I suspect that these values only really make sense when using a `cudaArray_t`, since the element size is never specified.

Reply:
I mean, to me the intended use of the 3D memcpys is kind of nebulous anyways because they accept more information than is actually needed. Ultimately, if the code works it doesn't matter. So as far as I'm concerned this PR can be merged regardless of whether or not this is changed.